### ISSUE TITLE
[ASCollectionView] AsyncDisplayKit w/ self-sizing UICollectionView layouts?

### ISSUE THREAD

**Issue by fatuhoku** [facebook/AsyncDisplayKit#973]
I'm trying to combine the layout of https://github.com/adamwaite/SelfSizingWaterfallCollectionViewLayout with the performance characteristics of `SelfSizingWaterfallCollectionViewLayout`.

There's the steps I took as a first stab:
1. Copied the `ASCollectionView` Sample project
2. Replace mentions of `UICollectionViewFlowLayout` with `SelfSizingWaterfallCollectionViewLayout`; do minimal configuration
3. Run. Error! ASDK told me that I need to set `layoutDelegate` (actually called `layoutInspector`).
4. I fix the error by copying `ASCollectionViewFlowLayoutInspector`'s delegate method implementations into the `ViewController`, and setting `_collectionView.layoutInspector = self;`. I was careful to remove any cases for horizontal scrolling from delegate methods (Waterfalls scroll vertically only!)
5. Run. Problem! **Observe that cells are no longer drawn!**

Why? A straight replacement of layout object should not produce this behaviour.  The only thing I can think of is that for whatever reason, `SelfSizingWaterfallCollectionViewLayout` is expecting cells (... however they're produced) to size themselves. However, no cells are present.

I can verify that:

1. `collectionView:nodeForItemAtIndexPath:` gets run as many times as there are cells (100 sections x 10 cells each = 1000 cells in total)
2. `collectionView:nodeForSupplementaryElementOfKind:atIndexPath:` gets run as many times as there are sections (100 sections)

The layout process seems to be really CPU-bound as well. When I first run the app, it looks like this:

![simulator screen shot 22 dec 2015 16 07 58](https://cloud.githubusercontent.com/assets/430233/11959620/44710fa8-a8c6-11e5-89a6-b1c75075c208.png)

... and after 6 seconds, the layout completes and looks like this (notice that cells aren't drawn):

![simulator screen shot 22 dec 2015 16 08 05](https://cloud.githubusercontent.com/assets/430233/11959638/59e55858-a8c6-11e5-9b04-4e5dfa56d3b9.png)

I'm a newbie at ASDK, so I'd really appreciate it if anybody can tell me just what's going on when I make these very simple replacements.  I really don't understand the sequence and interactions between ASDK and UIKit, especially with regards to self sizing and `preferredLayoutAttributesFittingAttributes:`

Cheers.

**Comment by fatuhoku** [comment]
I've uploaded the exact code I'm running here: https://github.com/fatuhoku/asdk-self-sizing-collection-code-sample/tree/add-self-sizing-waterfall-layout

**Comment by nguyenhuy** [comment]
After a quick debugging, it turns out that you need to have a separate inspector class, because ASCollectionViewLayoutInspecting and ASCollectionViewDataSource protocols declare `collectionView:constrainedSizeForNodeAtIndexPath:`. So if a class conforms to both protocol and provides an zero constrained size, that size will be picked up by ASCollectionView ([here](https://github.com/facebook/AsyncDisplayKit/blob/937b72b64a01d1735052e8b49acf73959b4d3844/AsyncDisplayKit/ASCollectionView.mm#L671)) and thus all cell nodes will be measured using the invalid size.

@levi Is there anything we can do to avoid this kind of mistake? Like having ASCollectionViewLayoutInspecting to require a method with slightly different signature?

@fatuhoku You may want to take a look at the MosaicCollectionViewLayout in CustomCollectionView example. It produces a Pinterest-like layout that is pretty close to what you are trying to achieve with SelfSizingWaterfallCollectionViewLayout. For difficulties building a custom collection view layout, definitely ask @levi. He is the expert :)

**Comment by levi** [comment]
@fatuhoku you'll want to check out the CustomCollectionViewLayoir example project to see how to use a custom layout with a layout inspector. The inspector's job is to provide answers to the data controller about the underlying behavior of the collection view layout under specific conditions. The example project implements a layout with a waterfall-like grid system--it should give you a good place to start.

**Comment by fatuhoku** [comment]
@nguyenhuy oh cool, thanks for the intro!
@levi okay, I'll take a look soon and will let you know what I find.

**Comment by fatuhoku** [comment]
Okay @levi, I think this is best marked as a duplicate of #954 — I need the cells to size themselves, without the need for an external mechanism. The inspector sorts of ends up having to do size calculations outside of the `ASLayout` and `ASLayoutSpec` mechanism.  This causes duplication of layout logic and is no better than UIKit currently.

The real question is, is there a way to do this without necessarily re-implementing what UIKit introduced in the form of `estimatedItemSize`?

**Comment by levi** [comment]
@fatuhoku the cells do in fact size themselves, the layout inspector class only provides the constrained size for the cells to measure within. The layoutSpecThatFits: functional is still very much available for your supplementary nodes.

Maybe you could paste what you're looking to accomplish and I can help you define what a custom layout inspector looks like?

**Comment by fatuhoku** [comment]
@levi The `CustomCollectionView` example doesn't actually demonstrate self-sizing, because:

1. In `MosaicCollectionViewLayout`, it requires relies on a delegate object to get the image's original size, so that the height can be calculated from the fixed cell width and aspect ratio.
2. `ImageCellNode` assumes that the only contents of the cell is an `ImageNode` and nothing else. If we introduce, say, a `ASStackLayoutSpec` that stacks an image on top of text in there and run the example, MosaicCollectionViewLayout actually knows nothing about the newly added `ASTextNode`, and therefore fails to lay things out properly.

I expect that ASDK should use the cell's layout specs to drive the `UICollectionViewLayout`.  This simply isn't done at the moment, and I think it still occurs on the main thread. Am I right?

Basically I'm looking to create more or less the behaviour of the infinite scrolling home page of Pinterest — each item has an image (with its own aspect ratio), its own description and caption. These nodes are stacked together to give the height of the cell.

**Comment by levi** [comment]
@fatuhoku see my reply on #954, as it addresses the points you bring up here.

**Comment by fatuhoku** [comment]
I managed to resolve it thanks for @vtsoup's custom demo project over at #954.
